consensus/hotstuff/votecollector/statemachine_test.go

```diff
@@ -0,0 +1,200 @@
+	block := helper.MakeBlock(s.T(), helper.WithBlockView(s.view))
+	proposal := helper.MakeProposal(s.T(), helper.WithBlock(block))
+	processor := s.prepareMockedProcessor(block)
+	err := s.collector.ProcessBlock(proposal)
+	require.NoError(s.T(), err)
+	s.T().Run("add-valid-vote", func(t *testing.T) {
+		vote := unittest.VoteForBlockFixture(block)
+		processor.On("Process", vote).Return(nil).Once()
+		err := s.collector.AddVote(vote)
+		require.NoError(t, err)
+		processor.AssertCalled(t, "Process", vote)
+	})
+	s.T().Run("add-double-vote", func(t *testing.T) {
+		firstVote := unittest.VoteForBlockFixture(block)
+		processor.On("Process", firstVote).Return(nil).Once()
+		err := s.collector.AddVote(firstVote)
+		require.NoError(t, err)
+
+		secondVote := unittest.VoteFixture(func(vote *model.Vote) {
+			vote.View = firstVote.View
+			vote.SignerID = firstVote.SignerID
+		})
+
+		s.notifier.On("OnDoubleVotingDetected", firstVote, secondVote).Return(nil).Once()
+
+		err = s.collector.AddVote(secondVote)
+		// we shouldn't get an error
+		require.NoError(t, err)
+
+		// but should get notified about double voting
+		s.notifier.AssertCalled(t, "OnDoubleVotingDetected", firstVote, secondVote)
+		processor.AssertCalled(t, "Process", firstVote)
+	})
+	s.T().Run("add-invalid-vote", func(t *testing.T) {
+		vote := unittest.VoteForBlockFixture(block, unittest.WithVoteView(s.view))
+		processor.On("Process", vote).Return(model.NewInvalidVoteErrorf(vote, "")).Once()
+
+		s.notifier.On("OnInvalidVoteDetected", vote).Return(nil).Once()
+		err := s.collector.AddVote(vote)
+		// in case process returns model.InvalidVoteError we should silently ignore this error
+		require.NoError(t, err)
+
+		// but should get notified about invalid vote
+		s.notifier.AssertCalled(t, "OnInvalidVoteDetected", vote)
+		processor.AssertCalled(t, "Process", vote)
+	})
+	s.T().Run("add-repeated-vote", func(t *testing.T) {
+		vote := unittest.VoteForBlockFixture(block)
+		processor.On("Process", vote).Return(nil).Once()
+		err := s.collector.AddVote(vote)
+		require.NoError(t, err)
+
+		// calling with same vote should exit early without error and don't do any extra processing
+		err = s.collector.AddVote(vote)
+		require.NoError(t, err)
+
+		processor.AssertCalled(t, "Process", vote)
+	})
+	s.T().Run("add-incompatible-view-vote", func(t *testing.T) {
+		vote := unittest.VoteForBlockFixture(block, unittest.WithVoteView(s.view+1))
+		err := s.collector.AddVote(vote)
+		require.ErrorIs(t, err, VoteForIncompatibleViewError)
+	})
+	s.T().Run("add-incompatible-block-vote", func(t *testing.T) {
+		vote := unittest.VoteForBlockFixture(block, unittest.WithVoteView(s.view))
+		processor.On("Process", vote).Return(VoteForIncompatibleBlockError).Once()
+		err := s.collector.AddVote(vote)
+		// in case process returns VoteForIncompatibleBlockError we should silently ignore this error
+		require.NoError(t, err)
+		processor.AssertCalled(t, "Process", vote)
+	})
+}
```

Comment:
we could also test that unexpected Processor errors are passed up the call stack:
```suggestion
	s.T().Run("unexpected-VoteProcessor-errors-are-passed-up", func(t *testing.T) {
		unexpectedError := errors.New("some unexpected error")
		vote := unittest.VoteForBlockFixture(block, unittest.WithVoteView(s.view))
		processor.On("Process", vote).Return(unexpectedError).Once()
		err := s.collector.AddVote(vote)
		require.ErrorIs(t, err, unexpectedError)
	})
}
```